### DISCUSSION TITLE
How to correctly specify a return type of getLogs?

### DISCUSSION THREAD

**Discussion by benesjan** [wevm/viem#359] [Question] [answered]
Hello, how would I specify a return type of getLogs?
<img width="809" alt="image" src="https://user-images.githubusercontent.com/13470840/231426967-9931482e-de90-43d9-bf4f-9efa386929a5.png">

**Comment by fubhy** [comment]
Normally, you shouldn't need to explicitly specifiy generic type variables yourself. The return type should be automatically inferred from the given abi & client instance.

Can you create a reproduction (e.g. stackblitz.com) if that isn't working for you?

**Reply by fubhy** [reply]
I suspect that you are also assigning a type to `this.publicClient`. If you are doing that, and not doing it correctly (e.g. `PublicClient<TTransport, TChain>`, then the `getLogs` won't be able to infer the type for you.

**Reply by benesjan** [reply]
Hello, when I do replaced
```
    return await this.publicClient.getLogs({
      address: getAddress(this.rollupAddress.toString()),
      event: getAbiItem({
          abi: RollupAbi,
          name: 'L2BlockProcessed',
      }),
      fromBlock,
      toBlock,
    });
```

with

```
    const abiItem = getAbiItem({
      abi: RollupAbi,
      name: 'L2BlockProcessed',
    });
    return await this.publicClient.getLogs({
      address: getAddress(this.rollupAddress.toString()),
      event: abiItem,
      fromBlock,
      toBlock,
    });
```

the return type got inferred correctly. 

The abi is defined [here](https://github.com/AztecProtocol/aztec3-packages/blob/master/yarn-project/l1-contracts/src/viem-abis/RollupAbi.ts) and the code where we use it is [here](https://github.com/AztecProtocol/aztec3-packages/blob/9bb32c2f1189d376ee2865476abbb6c6ed3a4a60/yarn-project/archiver/src/archiver/archiver.ts#L166). Do you see something wrong with the abi? If not I am fine with the solution we've came up with. Thank you for your help

**Reply by tmm** [reply, marked as answer]
Looks like the issue is internally we aren't narrowing `event` (e.g. `TAbiEvent`) so `event` defined inline doesn't work. Updated in this PR https://github.com/wagmi-dev/viem/pull/363

**Reply by fubhy** [reply]
~~Ok well that's a bug. And I can see where... I'll create a PR shortly.~~

Didn't see your message! Beat me to it :-)

**Reply by tmm** [reply]
Already fixed it here: https://github.com/wagmi-dev/viem/pull/363

**Reply by fubhy** [reply]
Beat me to it!
